File: golfclient/src/terrain.ts
export interface CourseTheme {
  skyTop: string       // hex — sky gradient top
  skyBottom: string    // hex — sky gradient bottom
  mountain1: string    // hex — far/back mountains
  mountain1Y: number   // baseline Y as fraction of worldH (0=top, 1=bottom)
  mountain2: string    // hex — near/front mountains
  mountain2Y: number   // baseline Y as fraction of worldH
  groundFill: string   // hex — terrain fill
  groundLine: string   // hex — terrain edge line
  groundLineW: number
  waterFill: string    // hex — water body fill
  waterLineW: number
  waterLine: string    // hex — water surface line
  sunColor: string     // hex — sun core
  sunRing1: string     // hex — inner glow ring
  sunRing2: string     // hex — outer glow ring
  sunSize: number      // px radius of sun core
}

export const DEFAULT_THEME: CourseTheme = {
  skyTop:     '#07071a',
  skyBottom:  '#111125',
  mountain1:  '#1c1c30',
  mountain1Y: 0.45,
  mountain2:  '#252540',
  mountain2Y: 0.58,
  groundFill: '#252515',
  groundLine: '#667755',
  groundLineW: 2,
  waterFill:  '#1655e1',
  waterLineW:  1.5,
  waterLine:  '#96d2ff',
  sunColor:   '#ffe9b8',
  sunRing1:   '#ffd27a',
  sunRing2:   '#ffd27a',
  sunSize:    32,
}

// Convert a #RRGGBB hex color to an rgba() string with the given alpha.
export function hexWithAlpha(hex: string, alpha: number): string {
  const r = parseInt(hex.slice(1, 3), 16)
  const g = parseInt(hex.slice(3, 5), 16)
  const b = parseInt(hex.slice(5, 7), 16)
  return `rgba(${r},${g},${b},${alpha})`
}

// ---- Spline terrain ----

export interface ControlPoint { x: number; y: number }

export interface SplineCoeff {
  x0: number; x1: number
  a: number; b: number; c: number; d: number
}

// Catmull-Rom spline. Phantom endpoints are reflected so the curve passes
// through the first and last control points without clamping.
// Neutral Base-Y reference for spline terrain. Splines are authored as absolute
// control-point Y values; treating BaseGround as an offset from this reference lets
// the Base Y slider raise/lower the whole spline (and spline+wave) terrain, just
// like it already does for waves. Equal to the default hole's baseGround so
// existing courses at the default render identically. Must match the server
// (terrain.SplineBaseRef in golfserver/terrain/terrain.go).
export const SPLINE_BASE_REF = 650

// Vertical offset that Base Y applies to all authored-absolute geometry — spline
// control points, bunker rims, and water level — so they shift together with the
// terrain when the Base Y slider moves. (Wave terrain already bakes in baseGround.)
export function baseOffset(h: { baseGround: number }): number {
  return h.baseGround - SPLINE_BASE_REF
}

// Bunker rim spline coefficients, shifted by the hole's Base-Y offset so the sand
// rim tracks the terrain. Used by every renderer + the physics build.
export function bunkerRimCoeffs(h: { baseGround: number }, topEdge: ControlPoint[]): SplineCoeff[] {
  const off = baseOffset(h)
  return buildSpline(topEdge.map((p) => ({ x: p.x, y: p.y + off })))
}

export function buildSpline(pts: ControlPoint[]): SplineCoeff[] {
  if (pts.length < 2) return []
  const p = [...pts].sort((a, b) => a.x - b.x)
  const n = p.length
  const out: SplineCoeff[] = []
  for (let i = 0; i < n - 1; i++) {
    const p0 = i > 0     ? p[i-1].y : 2*p[0].y   - p[1].y
    const p1 = p[i].y
    const p2 = p[i+1].y
    const p3 = i < n - 2 ? p[i+2].y : 2*p[n-1].y - p[n-2].y
    out.push({
      x0: p[i].x, x1: p[i+1].x,
      a: p1,
      b: 0.5 * (-p0 + p2),
      c: 0.5 * (2*p0 - 5*p1 + 4*p2 - p3),
      d: 0.5 * (-p0 + 3*p1 - 3*p2 + p3),
    })
  }
  return out
}

export function splineY(x: number, coeffs: SplineCoeff[]): number {
  if (coeffs.length === 0) return 650
  let s = coeffs[0]
  for (let i = 0; i < coeffs.length; i++) {
    s = coeffs[i]
    if (x <= s.x1 || i === coeffs.length - 1) break
  }
  const t = s.x1 === s.x0 ? 0 : (x - s.x0) / (s.x1 - s.x0)
  return s.a + t*(s.b + t*(s.c + t*s.d))
}

export function splineSlope(x: number, coeffs: SplineCoeff[]): number {
  if (coeffs.length === 0) return 0
  let s = coeffs[0]
  for (let i = 0; i < coeffs.length; i++) {
    s = coeffs[i]
    if (x <= s.x1 || i === coeffs.length - 1) break
  }
  const h = s.x1 - s.x0
  if (h === 0) return 0
  const t = (x - s.x0) / h
  return (s.b + t*(2*s.c + t*3*s.d)) / h
}

// ---- Wave terrain ----

export interface TerrainWave {
  amplitude: number
  period: number
  phase: number
}

export interface TerrainSegment {
  length: number
  waves: TerrainWave[]
}

export type HazardKind = 'water' | 'sand' | 'tree'  // 'sand'/'tree' are legacy; bunkers replace them

// ---- Bunkers ----

export interface Bunker {
  topEdge: Pt[]           // Catmull-Rom control points for the rim (sorted by x)
}

// ---- Platforms ----

export interface Pt { x: number; y: number }

export interface Platform {
  points: Pt[]
  zOrder: 'front' | 'back'  // front = above terrain, back = behind terrain
  fillColor: string
  edgeColor: string
}

// Signed area of a polygon in screen/Y-down coordinates.
// Positive → clockwise (CW) on screen, which is what NewEdge expects for
// correct outward normals (right-perpendicular of each edge direction).
export function polySignedArea(pts: Pt[]): number {
  let a = 0
  for (let i = 0; i < pts.length; i++) {
    const j = (i + 1) % pts.length
    a += pts[i].x * pts[j].y - pts[j].x * pts[i].y
  }
  return a / 2
}

// Returns pts in CW order (positive signed area) so physics normals are correct.
export function ensureCW(pts: Pt[]): Pt[] {
  return polySignedArea(pts) >= 0 ? pts : [...pts].reverse()
}

// Point-in-polygon test (ray casting).
export function pointInPoly(x: number, y: number, pts: Pt[]): boolean {
  let inside = false
  for (let i = 0, j = pts.length - 1; i < pts.length; j = i++) {
    const xi = pts[i].x, yi = pts[i].y, xj = pts[j].x, yj = pts[j].y
    if ((yi > y) !== (yj > y) && x < (xj - xi) * (y - yi) / (yj - yi) + xi) inside = !inside
  }
  return inside
}

// A hazard is positioned in world coordinates and lives on the Course so it's
// editable and saved/loaded with the map.
//
// Water is a flood-fill: the player shapes a valley in the terrain, drops an
// anchor at `cx` somewhere inside it, and `level` is the water-surface Y the
// valley fills up to. The body spreads left/right from the anchor until the
// ground rises above `level`, and fills between the terrain and the surface — so
// it always conforms to the land instead of being a floating box. `w`/`h` are
// unused for water; sand uses `w` (footprint) and `h` (mound height); trees use
// `h` (height).
export interface Hazard {
  kind: HazardKind
  cx: number
  w: number
  h: number
  level?: number
}

// Walks outward from `cx` to find where the terrain rises above `level` on
// each side — the pool's banks. Returns null if `cx` isn't actually underwater
// at `level` (ground there sits above the surface), which means the trap is
// misplaced and should render/collide as nothing rather than a phantom pool.
export function waterPoolBounds(
  cx: number,
  level: number,
  ptY: (x: number) => number,
  worldW: number,
  step = 4,
): { left: number; right: number } | null {
  if (ptY(cx) < level) return null

  let left = cx
  while (left - step >= 0 && ptY(left - step) >= level) left -= step
  if (left - step >= 0) {
    const yA = ptY(left), yB = ptY(left - step)
    left -= ((level - yA) / (yB - yA)) * step
  } else {
    left = 0
  }

  let right = cx
  while (right + step <= worldW && ptY(right + step) >= level) right += step
  if (right + step <= worldW) {
    const yA = ptY(right), yB = ptY(right + step)
    right += ((level - yA) / (yB - yA)) * step
  } else {
    right = worldW
  }

  return { left, right }
}

// A Hole is a single self-contained playable hole (formerly named Course — a
// "course" in the old single-hole code was really one hole). A Course is now an
// ordered list of Holes; rendering and physics are per-hole.
export interface Hole {
  name?: string
  par?: number
  worldW: number
  worldH: number
  baseGround: number
  teeBackX: number
  teeForwardX: number
  holeX: number
  useSpline: boolean
  controlPoints: ControlPoint[]
  useWaves: boolean
  segments: TerrainSegment[]
  hazards: Hazard[]
  bunkers: Bunker[]
  platforms: Platform[]
  theme: CourseTheme
}

// The format version this client authors. Migration of older files happens
// server-side at load time, so the client only ever receives current-format
// courses; this constant is what it stamps on new/edited courses it saves.
export const CURRENT_FORMAT_VERSION = 1

// A Course is the multi-hole unit stored as one file on the server.
export interface Course {
  formatVersion: number
  id: string
  name: string
  holes: Hole[]
}

// Precomputed segment — startX and offset are derived for height continuity at joints.
export interface BuiltSegment {
  startX: number
  length: number
  waves: TerrainWave[]
  offset: number  // vertical shift so terrainY(startX) == previous segment's end Y
}

// Build segments from a Course, computing startX and offset for each so adjacent
// segments always connect at exactly the same height (C0 continuity, slope may kink).
export function buildSegments(course: Hole): BuiltSegment[] {
  let curX = 0
  let curY = course.baseGround
  return course.segments.map(seg => {
    // At localX=0: waveSum = Σ(amp·sin(phase))
    const atStart = seg.waves.reduce((s, w) => s + w.amplitude * Math.sin(w.phase), 0)
    const offset = curY - atStart
    // End Y for the next segment's base
    const atEnd = seg.waves.reduce((s, w) =>
      s + w.amplitude * Math.sin(seg.length / w.period + w.phase), 0)
    curY = offset + atEnd
    const built: BuiltSegment = { startX: curX, length: seg.length, waves: seg.waves, offset }
    curX += seg.length
    return built
  })
}

export function terrainY(x: number, segs: BuiltSegment[]): number {
  for (let i = 0; i < segs.length; i++) {
    const s = segs[i]
    if (x < s.startX + s.length || i === segs.length - 1) {
      const lx = x - s.startX
      return s.offset + s.waves.reduce((acc, w) =>
        acc + w.amplitude * Math.sin(lx / w.period + w.phase), 0)
    }
  }
  return segs[0]?.offset ?? 650
}

export function terrainSlope(x: number, segs: BuiltSegment[]): number {
  for (let i = 0; i < segs.length; i++) {
    const s = segs[i]
    if (x < s.startX + s.length || i === segs.length - 1) {
      const lx = x - s.startX
      return s.waves.reduce((acc, w) =>
        acc + (w.amplitude / w.period) * Math.cos(lx / w.period + w.phase), 0)
    }
  }
  return 0
}

// Default hole: single segment replicating the original 3-sinusoid formula exactly.
export const DEFAULT_HOLE: Hole = {
  worldW: 4000,
  worldH: 1000,
  baseGround: 650,
  teeBackX: 200,
  teeForwardX: 400,
  holeX: 3700,
  useSpline: false,
  controlPoints: [
    { x: 0,    y: 650 },
    { x: 700,  y: 570 },
    { x: 1400, y: 710 },
    { x: 2100, y: 625 },
    { x: 2800, y: 695 },
    { x: 3500, y: 610 },
    { x: 4000, y: 650 },
  ],
  useWaves: true,
  segments: [{
    length: 4000,
    waves: [
      { amplitude: 80, period: 800, phase: 0 },
      { amplitude: 40, period: 300, phase: 0 },
      { amplitude: 20, period: 150, phase: 0 },
    ]
  }],
  hazards: [
    { kind: 'water', cx: 2080, w: 0, h: 0, level: 715 },
  ],
  bunkers: [],
  platforms: [],
  // Own copy, not the shared DEFAULT_THEME reference — otherwise a hole that ends
  // up aliasing this object could mutate the global default in place.
  theme: { ...DEFAULT_THEME },
}

File: golfserver/terrain/terrain.go
package terrain

import (
	"math"
	"sort"
)

// ---- Spline terrain ----

type ControlPoint struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type SplineCoeff struct {
	X0, X1    float64
	A, B, C, D float64
}

func BuildSpline(pts []ControlPoint) []SplineCoeff {
	if len(pts) < 2 {
		return nil
	}
	p := make([]ControlPoint, len(pts))
	copy(p, pts)
	sort.Slice(p, func(i, j int) bool { return p[i].X < p[j].X })
	n := len(p)
	out := make([]SplineCoeff, n-1)
	for i := 0; i < n-1; i++ {
		var p0, p3 float64
		p1, p2 := p[i].Y, p[i+1].Y
		if i > 0 {
			p0 = p[i-1].Y
		} else {
			p0 = 2*p[0].Y - p[1].Y
		}
		if i < n-2 {
			p3 = p[i+2].Y
		} else {
			p3 = 2*p[n-1].Y - p[n-2].Y
		}
		out[i] = SplineCoeff{
			X0: p[i].X, X1: p[i+1].X,
			A: p1,
			B: 0.5 * (-p0 + p2),
			C: 0.5 * (2*p0 - 5*p1 + 4*p2 - p3),
			D: 0.5 * (-p0 + 3*p1 - 3*p2 + p3),
		}
	}
	return out
}

func SplineY(x float64, coeffs []SplineCoeff) float64 {
	if len(coeffs) == 0 {
		return 650
	}
	s := coeffs[0]
	for i := range coeffs {
		s = coeffs[i]
		if x <= s.X1 || i == len(coeffs)-1 {
			break
		}
	}
	if s.X1 == s.X0 {
		return s.A
	}
	t := (x - s.X0) / (s.X1 - s.X0)
	return s.A + t*(s.B+t*(s.C+t*s.D))
}

func SplineSlope(x float64, coeffs []SplineCoeff) float64 {
	if len(coeffs) == 0 {
		return 0
	}
	s := coeffs[0]
	for i := range coeffs {
		s = coeffs[i]
		if x <= s.X1 || i == len(coeffs)-1 {
			break
		}
	}
	h := s.X1 - s.X0
	if h == 0 {
		return 0
	}
	t := (x - s.X0) / h
	return (s.B + t*(2*s.C+t*3*s.D)) / h
}

// ComputeTerrainY and ComputeTerrainSlope handle all four mode combinations.
// SplineBaseRef is the neutral Base-Y reference for spline terrain — see the TS
// SPLINE_BASE_REF in golfclient/src/terrain.ts (must match). BaseGround acts as an
// offset from this so the Base Y slider shifts the whole spline terrain.
const SplineBaseRef = 650.0

// BaseOffset is the vertical shift Base Y applies to authored-absolute geometry
// (spline control points, bunker rims, water level) so they track the terrain.
func BaseOffset(c Hole) float64 { return c.BaseGround - SplineBaseRef }

// BunkerRimCoeffs builds a bunker's rim spline shifted by the hole's Base-Y offset.
func BunkerRimCoeffs(c Hole, topEdge []ControlPoint) []SplineCoeff {
	off := BaseOffset(c)
	shifted := make([]ControlPoint, len(topEdge))
	for i, p := range topEdge {
		shifted[i] = ControlPoint{X: p.X, Y: p.Y + off}
	}
	return BuildSpline(shifted)
}

func ComputeTerrainY(x float64, c Hole, segs []BuiltSegment, coeffs []SplineCoeff) float64 {
	s, w := c.UseSpline, c.UseWaves
	if s && w {
		return SplineY(x, coeffs) + TerrainY(x, segs) - SplineBaseRef
	}
	if s {
		return SplineY(x, coeffs) + c.BaseGround - SplineBaseRef
	}
	if w {
		return TerrainY(x, segs)
	}
	return c.BaseGround
}

func ComputeTerrainSlope(x float64, c Hole, segs []BuiltSegment, coeffs []SplineCoeff) float64 {
	s, w := c.UseSpline, c.UseWaves
	if s && w {
		return SplineSlope(x, coeffs) + TerrainSlope(x, segs)
	}
	if s {
		return SplineSlope(x, coeffs)
	}
	if w {
		return TerrainSlope(x, segs)
	}
	return 0
}

// ---- Theme ----

type CourseTheme struct {
	SkyTop      string  `json:"skyTop"`
	SkyBottom   string  `json:"skyBottom"`
	Mountain1   string  `json:"mountain1"`
	Mountain1Y  float64 `json:"mountain1Y"`
	Mountain2   string  `json:"mountain2"`
	Mountain2Y  float64 `json:"mountain2Y"`
	GroundFill  string  `json:"groundFill"`
	GroundLine  string  `json:"groundLine"`
	GroundLineW float64 `json:"groundLineW"`
	WaterFill   string  `json:"waterFill"`
	WaterLineW  float64 `json:"waterLineW"`
	WaterLine   string  `json:"waterLine"`
	SunColor    string  `json:"sunColor"`
	SunRing1    string  `json:"sunRing1"`
	SunRing2    string  `json:"sunRing2"`
	SunSize     float64 `json:"sunSize"`
}

type TerrainWave struct {
	Amplitude float64 `json:"amplitude"`
	Period    float64 `json:"period"`
	Phase     float64 `json:"phase"`
}

type TerrainSegment struct {
	Length float64       `json:"length"`
	Waves  []TerrainWave `json:"waves"`
}

// Bunker is a sand trap whose visible region sits between a user-drawn
// Catmull-Rom top edge (the rim/lip) and the terrain surface below it. Ball
// physics inside a bunker (extra rolling friction, reduced shot power when hit
// out) is governed by global tunables — physics.Current.BunkerFriction and the
// per-club penalties — rather than per-bunker fields, so every bunker on every
// hole behaves the same.
type Bunker struct {
	TopEdge []ControlPoint `json:"topEdge"`
}

// Platform is a convex or concave polygon that the ball can bounce off.
// ZOrder controls rendering: "front" draws above terrain, "back" draws behind it.
// Points should be stored in the order the editor produced them; EnsureCW
// normalises winding for physics before building edges.
type Platform struct {
	Points    []ControlPoint `json:"points"`
	ZOrder    string         `json:"zOrder"` // "front" | "back"
	FillColor string         `json:"fillColor"`
	EdgeColor string         `json:"edgeColor"`
}

// PolySignedArea returns the signed area of a polygon in screen/Y-down
// coordinates. Positive → clockwise (CW) on screen.
func PolySignedArea(pts []ControlPoint) float64 {
	a := 0.0
	n := len(pts)
	for i := 0; i < n; i++ {
		j := (i + 1) % n
		a += pts[i].X*pts[j].Y - pts[j].X*pts[i].Y
	}
	return a / 2
}

// EnsureCW returns pts in CW winding order (positive signed area in screen
// coords) so physics.NewEdge outward normals point away from the platform.
func EnsureCW(pts []ControlPoint) []ControlPoint {
	if PolySignedArea(pts) >= 0 {
		return pts
	}
	rev := make([]ControlPoint, len(pts))
	for i, p := range pts {
		rev[len(pts)-1-i] = p
	}
	return rev
}

// Hazard is positioned and sized in world coordinates. Water is a flood-fill:
// the body floods outward from the anchor (CX) until the terrain rises above
// Level, conforming to whatever valley it sits in (see WaterPoolBounds); W/H
// are unused for water. Sand uses W (footprint) and H (mound height); trees
// use H (height).
type Hazard struct {
	Kind  string  `json:"kind"`
	CX    float64 `json:"cx"`
	W     float64 `json:"w"`
	H     float64 `json:"h"`
	Level float64 `json:"level"`
}

// WaterPoolBounds walks outward from cx to find where the terrain rises above
// level on each side — the pool's banks. ok is false if cx isn't actually
// underwater at level (ground there sits above the surface), meaning the trap
// is misplaced and should be treated as having no pool.
func WaterPoolBounds(cx, level float64, cty func(float64) float64, worldW float64) (left, right float64, ok bool) {
	const step = 4.0
	if cty(cx) < level {
		return 0, 0, false
	}

	left = cx
	for left-step >= 0 && cty(left-step) >= level {
		left -= step
	}
	if left-step >= 0 {
		yA, yB := cty(left), cty(left-step)
		left -= ((level - yA) / (yB - yA)) * step
	} else {
		left = 0
	}

	right = cx
	for right+step <= worldW && cty(right+step) >= level {
		right += step
	}
	if right+step <= worldW {
		yA, yB := cty(right), cty(right+step)
		right += ((level - yA) / (yB - yA)) * step
	} else {
		right = worldW
	}

	return left, right, true
}

// Hole is a single self-contained playable hole: terrain, tees, cup, hazards,
// and theme. It was formerly named Course — a "course" in the old code was
// really one hole. A full Course (see package coursestore) is now an ordered
// list of Holes. Everything downstream — physics edge-building, rendering — is
// per-hole, so this struct is the unit sent to the physics/collision layer.
type Hole struct {
	Name          string           `json:"name,omitempty"`
	Par           int              `json:"par,omitempty"`
	WorldW        float64          `json:"worldW"`
	WorldH        float64          `json:"worldH"`
	BaseGround    float64          `json:"baseGround"`
	TeeBackX      float64          `json:"teeBackX"`
	TeeForwardX   float64          `json:"teeForwardX"`
	HoleX         float64          `json:"holeX"`
	UseSpline     bool             `json:"useSpline"`
	ControlPoints []ControlPoint   `json:"controlPoints"`
	UseWaves      bool             `json:"useWaves"`
	Segments      []TerrainSegment `json:"segments"`
	Hazards       []Hazard         `json:"hazards"`
	Bunkers       []Bunker         `json:"bunkers"`
	Platforms     []Platform       `json:"platforms"`
	Theme         CourseTheme      `json:"theme"`
}

// BuiltSegment is a precomputed segment with continuity offset.
type BuiltSegment struct {
	StartX float64
	Length float64
	Waves  []TerrainWave
	Offset float64 // vertical shift so Y(StartX) == previous segment's end Y
}

// BuildSegments computes startX and offset for each segment so adjacent
// segments always connect at the same height (C0 continuity).
func BuildSegments(c Hole) []BuiltSegment {
	built := make([]BuiltSegment, len(c.Segments))
	curX := 0.0
	curY := c.BaseGround
	for i, seg := range c.Segments {
		atStart := 0.0
		for _, w := range seg.Waves {
			atStart += w.Amplitude * math.Sin(w.Phase)
		}
		offset := curY - atStart
		atEnd := 0.0
		for _, w := range seg.Waves {
			atEnd += w.Amplitude * math.Sin(seg.Length/w.Period+w.Phase)
		}
		curY = offset + atEnd
		built[i] = BuiltSegment{StartX: curX, Length: seg.Length, Waves: seg.Waves, Offset: offset}
		curX += seg.Length
	}
	return built
}

func TerrainY(x float64, segs []BuiltSegment) float64 {
	for i, s := range segs {
		if x < s.StartX+s.Length || i == len(segs)-1 {
			lx := x - s.StartX
			y := s.Offset
			for _, w := range s.Waves {
				y += w.Amplitude * math.Sin(lx/w.Period+w.Phase)
			}
			return y
		}
	}
	if len(segs) > 0 {
		return segs[0].Offset
	}
	return 650
}

func TerrainSlope(x float64, segs []BuiltSegment) float64 {
	for i, s := range segs {
		if x < s.StartX+s.Length || i == len(segs)-1 {
			lx := x - s.StartX
			slope := 0.0
			for _, w := range s.Waves {
				slope += (w.Amplitude / w.Period) * math.Cos(lx/w.Period+w.Phase)
			}
			return slope
		}
	}
	return 0
}

// DefaultHole returns the original 3-sinusoid terrain as a single segment.
func DefaultHole() Hole {
	return Hole{
		WorldW:      4000,
		WorldH:      1000,
		BaseGround:  650,
		TeeBackX:    200,
		TeeForwardX: 400,
		HoleX:       3700,
		UseSpline:   false,
		ControlPoints: []ControlPoint{
			{X: 0, Y: 650}, {X: 700, Y: 570}, {X: 1400, Y: 710},
			{X: 2100, Y: 625}, {X: 2800, Y: 695}, {X: 3500, Y: 610}, {X: 4000, Y: 650},
		},
		UseWaves: true,
		Segments: []TerrainSegment{{
			Length: 4000,
			Waves: []TerrainWave{
				{Amplitude: 80, Period: 800, Phase: 0},
				{Amplitude: 40, Period: 300, Phase: 0},
				{Amplitude: 20, Period: 150, Phase: 0},
			},
		}},
		Hazards:   []Hazard{{Kind: "water", CX: 2080, Level: 715}},
		Bunkers:   []Bunker{},
		Platforms: []Platform{},
		Theme: CourseTheme{
			SkyTop:      "#07071a",
			SkyBottom:   "#111125",
			Mountain1:   "#1c1c30",
			Mountain1Y:  0.45,
			Mountain2:   "#252540",
			Mountain2Y:  0.58,
			GroundFill:  "#252515",
			GroundLine:  "#667755",
			GroundLineW: 2,
			WaterFill:   "#1655e1",
			WaterLineW:  1.5,
			WaterLine:   "#96d2ff",
			SunColor:    "#ffe9b8",
			SunRing1:    "#ffd27a",
			SunRing2:    "#ffd27a",
			SunSize:     32,
		},
	}
}
